de.peeeq.wurstscript/src/main/java/de/peeeq/wurstscript/intermediatelang/optimizer/BranchMerger.java

```diff
@@ -0,0 +1,109 @@
+        mergeBranches(func);
+    }
+
+
+    private void mergeBranches(ImFunction func) {
+        ControlFlowGraph cfg = new ControlFlowGraph(func.getBody());
+
+        // init in and out with empty sets
+        for (Node n : cfg.getNodes()) {
+            @Nullable ImStmt stmt = n.getStmt();
+            if (stmt != null && stmt.getParent() != null && stmt.getParent() instanceof ImIf) {
+                ImIf imIf = (ImIf) stmt.getParent();
+                if (n.getPredecessors().size() <= 1 && n.getSuccessors().size() == 2) {
+                    Node leftStmt = n.getSuccessors().get(0);
+                    Node rightStmt = n.getSuccessors().get(1);
+                    if (leftStmt.toString().equals(rightStmt.toString())) {
+                        if (n.getPredecessors().size() == 1) {
+                            // Possible match. At last check if condition causes sideeffects.
+                            if (hasNoSideEffects(n, leftStmt)) {
+
+                                ImStmt mergedStmt = leftStmt.getStmt().copy();
+                                ImIf oldIf = imIf.copy();
+                                oldIf.getThenBlock().get(0).replaceBy(JassIm.ImNull());
+                                oldIf.getElseBlock().get(0).replaceBy(JassIm.ImNull());
+                                imIf.replaceBy(JassIm.ImStatementExpr(JassIm.ImStmts(mergedStmt, oldIf), JassIm.ImNull()));
+                            }
+                        }
+                    }
+                }
+                // if the first statement of each branch of the if is the same, it can be moved before the branch
+            }
+        }
+    }
+
+
+    private List<ImVar> calculateVarUses(Node node) {
+        List<ImVar> result = Lists.newArrayList();
+        ImStmt stmt = node.getStmt();
+        if (stmt != null) {
+            stmt.accept(new ImStmt.DefaultVisitor() {
+                @Override
+                public void visit(ImFunctionCall call) {
+                    super.visit(call);
+                    visit(call.getFunc());
+                }
+
+                @Override
+                public void visit(ImVarAccess va) {
+                    super.visit(va);
+                    result.add(va.getVar());
+                }
+
+                @Override
+                public void visit(ImVarArrayAccess va) {
+                    super.visit(va);
+                    result.add(va.getVar());
+                }
+
+                @Override
+                public void visit(ImSet va) {
+                    super.visit(va);
+                    result.add(va.getLeft());
+                }
+            });
+        }
+        return result;
+    }
+
+    private boolean hasNoSideEffects(Node ifNode, Node stmtNode) {
+        List<ImVar> usesIf = calculateVarUses(ifNode);
+        List<ImVar> usesStmt = calculateVarUses(stmtNode);
+
+        usesIf.retainAll(usesStmt);
+        return usesIf.size() == 0;
```

Comment:
`hasNoSideEffects` should be named `sideEffectsCanAffectNode` or something like that: Or add a comment that you are checking, if executing `stmtNode` could affect the condition of the if-statement.